src/main/java/io/r2dbc/pool/ConnectionPool.java

```diff
@@ -39,34 +39,36 @@
 
     private final Pool<Connection> connectionPool;
 
+    private final ConnectionPoolConfiguration configuration;
+
     /**
      * Creates a new connection factory.
      *
      * @param configuration the configuration to use connections
      * @throws IllegalArgumentException if {@code configuration} is {@code null}
+     * @throws IllegalArgumentException if {@link ConnectionPoolConfiguration#getConnectionFactory()} or {@link ConnectionPoolConfiguration#getCustomizer()} is {@code null}
      */
     public ConnectionPool(ConnectionPoolConfiguration configuration) {
-        this(Assert.requireNonNull(configuration.getConnectionFactory(), "ConnectionPoolConfiguration must not be null"), configuration.getCustomizer());
-    }
+        Assert.requireNonNull(configuration, "ConnectionPoolConfiguration must not be null");
+        Assert.requireNonNull(configuration.getConnectionFactory(), "ConnectionFactory must not be null");
+        Assert.requireNonNull(configuration.getCustomizer(), "poolBuilderCustomizer must not be null");
 
-    /**
-     * Creates a new connection factory.
-     *
-     * @param factory               the {@link ConnectionFactory} that creates actual database connections
-     * @param poolBuilderCustomizer {@link Consumer customizer} for the {@link PoolBuilder pool builder}.
-     * @throws IllegalArgumentException if {@code factory} or {@code poolBuilderCustomizer} is {@code null}
-     */
-    public ConnectionPool(ConnectionFactory factory, Consumer<PoolBuilder<Connection>> poolBuilderCustomizer) {
```

Comment:
Let's keep the `Consumer`-based configuration. This gives a lot of flexibility to users.

Reply:
I put back the constructor. Since timeout adds two new params(`maxCreateConnectionTime` and `maxAcquireTime`), I think this constructor gets more params when configurations are needed on `ConnectionPool` class.
If params grow more, I think it is better to keep the `ConnectionPoolConfiguration` as a local variable, and derive all parameters from it including the `Consumer` as I did in my first review.

Reply:
I'm not sure regarding both arguments in this configuration variant. It's quite simple to override `ConnectionFactory` adding the timeout operator to `create()` and on the other hand, one can add timeout operators to `ConnectionPool.create().timeout(…)`.

I'm inclined to keep construction via `ConnectionPool(ConnectionFactory, Consumer)` as it was, without a timeout configuration. This constructor aims for a lot of control over the pooling configuration and timeouts are a convenience function that is added if you configure the pool using a configuration object.

We have already a similar feature with connection validation, that is provided through `ConnectionPoolConfiguration`.

Does this make sense?

Reply:
Hmm, not quite clear about it.

So, there are two timeouts, `maxCreateConnectionTime` and `maxAcquireTime`.

For `maxCreateConnectionTime`, this needs to be applied to the publisher that `connectionPool` uses(used by `PoolBuilder` constructor)(line 83 `allocator` instance on ConnectionPool.java in my patch).
In order to `maxCreateConnectionTime` to be applied, it needs to be accessible before creating the pool which currently is happening in the second constructor with customizer.
Since customizer is applied after the `connectionPool` has created, timeout cannot be applied in customizer. It is late in customizer.

I think problem is that the first constructor(with configuration) calls the second constructor(with customizer) just by passing parameters instead of configuration. So that any other parameters in configuration is not accessible in the second constructor. Then, in the second constructor, it creates the `connectionPool` object which requires `maxCreateConnectionTime` parameter to be accessible.


For `maxAcquireTime`, it can be applied after `ConnectionPool` is created. `new ConnectionPool(...).acquireTime(...)`. 

`maxAcquireTime` is used in the Mono that was created in `ConnectionPool#create`.
Since the evaluation of this mono is at execution time, `maxAcquireTime` can be stored in instance variable and be evaluated lazily, so can be overriden after creation of `ConnectionPool`.
However, I think it is not intuitive and feels different approach than any other place.
So, I prefer the value to be determined at initial creation time via configuration object, rather than updating it lazily.

----

OR, something like this:

```java
private final ConnectionFactory factory;
private final Pool<Connection> connectionPool;
private Duration maxAcquireTime;
private Duration maxCreateConnectionTime;

public ConnectionPool(ConnectionPoolConfiguration configuration) {
    this.factory = configuration.getConnectionFactory();
    this.maxAcquireTime = configuration.getMaxAcquireTime();
    this.maxCreateConnectionTime = configuration.getMaxCreateConnectionTime();

    this.connectionPool = initializeConnectionPool(configuration.getCustomizer());
}

public ConnectionPool(ConnectionFactory factory, Consumer<PoolBuilder<Connection>> poolBuilderCustomizer) {
    this.factory = factory;
    this.maxAcquireTime = Duration.ZERO;
    this.maxCreateConnectionTime = Duration.ZERO;

    this.connectionPool = initializeConnectionPool(poolBuilderCustomizer);
}

private Pool<Connection> initializeConnectionPool(Consumer<PoolBuilder<Connection>> poolBuilderCustomizer) {

    // create PoolBuilder with applying "maxCreateConnectionTime"
    PoolBuilder<Connection> builder = ...

    poolBuilderCustomizer.accept(builder);
    return builder.build();
}
```

In this case, `maxCreateConnectionTime` cannot be configured if constructor with customizer is used.


I agree the flexibility of customizer. 

But how about making `ConnectionPoolConfiguration` to be a centralized configuration object and give setter/getter of customizer to be public?? So that, factory and customizer can be obtained from configuration and can also give flexibility to customize them by others?

I think constructor with factory and customizer will limit the ability to apply timeout.

Reply:
and `maxAcquireTime` can be configured in customizer if it becomes configurable in reactor-pool side :)

Reply:
Adding a `Consumer<PoolBuilder<Connection>>` to `ConnectionPoolConfiguration` seems a reasonable approach to retain flexibility and eliminating the issues that would come with a constructor accepting just `ConnectionFactory` and `Consumer<PoolBuilder<Connection>>`.

So we can get back to a single constructor accepting `ConnectionPoolConfiguration` and internally we apply `Consumer<PoolBuilder<Connection>>` and additonally, if configured, another `Consumer<PoolBuilder<Connection>>` that was configured on  `ConnectionPoolConfiguration`.

We should keep `ConnectionPoolConfiguration` immutable and keep visibility of getter methods at a minimum.

Reply:
ok, added a new commit, that does:

- Changed `ConnectionPool` to have single constructor that takes configuration
- Renamed `getCustomizer()` to `getInitialCustomizer()` to perform existing customization logic.
- Added `getCustomizer()` to return additional customizer that user can add on

In addition, we could move the existing customizer logic to `ConnectionPool` since there are two customizers in configuration now (one for default setup and one for additional setup). If we move the logic, `ConnectionPoolConfiguration` becomes an immutable value object.

`ConnectionPool` constructor will become:

```java

public ConnectionPool(ConnectionPoolConfiguration configuration) {
    // ... null check ...

    // values from configuration
    ConnectionFactory factory = configuration.getConnectionFactory();
    Duration maxCreateConnectionTime = configuration.getMaxCreateConnectionTime();
    Duration maxAcquireTime = configuration.getMaxAcquireTime();
    int initialSize = configuration.getInitialSize();
    int maxSize = configuration.getMaxSize();
    String validationQuery = configuration.getValidationQuery();
    Duration maxIdleTime = configuration.getMaxIdleTime();
    Consumer<PoolBuilder<Connection>> customizer = configuration.getCustomizer();

    // set timeout for create connection
    Mono<Connection> allocator = Mono.from(factory.create());
    if (!maxCreateConnectionTime.isZero()) {
        allocator = allocator.timeout(maxCreateConnectionTime);
    }

    PoolBuilder<Connection> builder = PoolBuilder.from(allocator)
            .destroyHandler(Connection::close)
            .sizeMax(Runtime.getRuntime().availableProcessors());

    //
    //  START logic from ConnectionPoolConfiguration#getInitialCustomizer()
    //
    builder.initialSize(initialSize);

    if (maxSize == -1) {
        builder.sizeUnbounded();
    } else {
        builder.sizeMax(maxSize);
    }

    if (validationQuery != null) {
        builder.releaseHandler(connection -> {
            return Flux.from(connection.createStatement(validationQuery).execute()).flatMap(it -> it.map((row, rowMetadata) -> Optional.ofNullable(row.get(0)))).then();
        });
    }

    builder.evictionIdle(maxIdleTime);
    //
    //  END logic from ConnectionPoolConfiguration#getInitialCustomizer()
    //

    // call additional customizer (default is no-op)
    customizer.accept(builder);

    this.connectionPool = builder.build();
    this.factory = factory;
    this.maxAcquireTime = maxAcquireTime;
}
```

This looks cleaner to me :)
Let me know, if we want to move the connection-pool construction logic or not.